Implement deep diffing of KeePass Databases

sseemayer opened the pull request:
First of all, thanks for this super-useful tool! I am using Dropbox to sync my KeePass databases and have been running in conflict situations where this tool is very helpful.

Second, I deeply apologize for the density of this PR. I fully understand if you are not willing to merge  this large of a changeset from some random person dropping into your project :)

This implements a deep recursive diff of two KeePass databases down to the level of fields being added, removed or changed, and showing it as a nested tree that reflects the structure of the KeePass database:

![deepdiff](https://user-images.githubusercontent.com/736556/66717569-d5f88c00-edda-11e9-932b-2973f2455629.png)

A lot of the functionality has been packaged into the `Diff` trait which is implemented for the `Group` `Entry` and `Field` structs that correspond to the related objects in the `keepass` crate.

Narigo commented:
Hey @sseemayer you aren't a "random person" when you are maintaining the keepass crate we are using as a dependency!

I haven't come around looking into this PR yet (have been sick last week) so this is a heads up that I'll look into it next weekend, hope that's okay for you.
As I've written in the Readme, I'm pretty new to Rust and this is more or less the only project I'm using it right now. If you have an idea how I could add some automated tests to this project, I would be more than happy... 😉

sseemayer commented on src/diff/mod.rs:
```diff
@@ -1,103 +1,174 @@
+                }
+                write!(f, "{}+ {}\n", indent, right)
+            }
+        }?;
```
The [? operator](https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html) is the way to get the value within a `Result` type in a properly-error-handled way.

In this case, all of the match cases are returning a `std::fmt::Result` from the `write!` operation. If the operation succeeds with a return value of `Ok(())`, this gets ignored. If there is an `Err(some_error)`, this will get returned for the whole `diff_result_format` call's return value.

sseemayer replied:
Probably it would also make sense in this case to silently ignore any write errors and still attempt to always send the color reset below. Would you like me to adjust this?

sseemayer replied:
I added another commit to do so.

sseemayer commented on src/main.rs:
```diff
@@ -3,20 +3,20 @@ extern crate keepass;
+use diff::{group::Group, Diff, DiffDisplay};
+use keepass::{result::Error, result::Result, Database};
+
+#[allow(unused_imports)]
```
Whoops, this is a left-over from an earlier version. You are completely right. I will fix this when I am in front of my text editor again.

sseemayer commented:
I've added a bunch of commits addressing your review comments (you were absolutely right on all of them :smile: ). Regarding [rustfmt](https://github.com/rust-lang/rustfmt), it's the official Rust formatting tool that formats code to the official Rust style guide (it says to [use 4 spaces for indentation](https://doc.rust-lang.org/1.0.0/style/style/whitespace.html)). 

I personally use the official [rust plugin for vim](https://github.com/rust-lang/rust.vim) together with `let g:rustfmt_autosave = 1` in my config to automatically run rustfmt upon saving, but you can also trigger rustfmt from the command line as part of a build script.

sseemayer commented:
> If you have an idea how I could add some automated tests to this project, I would be more than happy...

My approach would be to also add a derive for `Eq` for `DiffResult`s. You could then write some tests that either build up the appropriate structure of `Group`/`Entry`/`Field` themselves and diff them, using `assert_eq!` to check for the expected result, or to simply use your already-existing test databases and `assert_eq!` to check for the overall differences between databases. I could take a stab at that once this PR is merged with a subsequent PR, if you would like me to.

Narigo commented:
Thanks again for your efforts! I've been pretty sick for the last week making it hard for me to look into Rust again... I would like to have another look into it by checking everything out, running it myself and creating a new screencast for it.

A few things for the roadmap though:
1. Automated Testing - I'd like to have unit tests but I think more important would be a check of the real CLI output against snapshots. If the latter is hard to do with Rust, maybe we can use a bash script for that...?
2. Add possibility to write the whole path of a field into the console instead of having the same path segments as `~` lines. So a single line for each change with the full path before it. I think right now, the recursive algorithm doesn't save this information. If we make it readily available, we might even create the possibility to auto-merge two conflicted files.
